upb/encode.c

```diff
@@ -0,0 +1,381 @@
+/* We encode backwards, to avoid pre-computing lengths (one-pass encode). */
+
+#include "upb/upb.h"
+#include "upb/encode.h"
+#include "upb/structs.int.h"
+
+#define UPB_PB_VARINT_MAX_LEN 10
+#define CHK(x) do { if (!(x)) { return false; } } while(0)
+
+/* Maps descriptor type -> upb field type.  */
+static const uint8_t upb_desctype_to_fieldtype[] = {
```

Comment:
I can't figure out what header file it should go into. I think I'll punt on this until there are more things of this sort, and it's more obvious where to put them.